src/rgw/rgw_sal.cc

```diff
@@ -236,28 +270,56 @@ int RGWRadosBucket::chown(RGWUser* new_user, RGWUser* old_user, optional_yield y
 			   old_user->get_display_name(), obj_marker, y);
 }
 
-bool RGWRadosBucket::is_owner(RGWUser* user)
+int RGWRadosBucket::put_instance_info(bool exclusive, ceph::real_time _mtime)
 {
-  get_bucket_info(null_yield);
+  mtime = _mtime;
+  return store->getRados()->put_bucket_instance_info(info, exclusive, mtime, &attrs.attrs);
+}
 
+/* Make sure to call get_bucket_info() if you need it first */
+bool RGWRadosBucket::is_owner(RGWUser* user)
+{
   return (info.owner.compare(user->get_user()) == 0);
 }
 
+int RGWRadosBucket::check_empty(optional_yield y)
+{
+  return store->getRados()->check_bucket_empty(info, y);
+}
+
+int RGWRadosBucket::check_quota(RGWQuotaInfo& user_quota, RGWQuotaInfo& bucket_quota, uint64_t obj_size)
+{
+    return store->getRados()->check_quota(owner->get_user(), get_bi(),
+					  user_quota, bucket_quota, obj_size);
+}
+
 int RGWRadosBucket::set_acl(RGWAccessControlPolicy &acl, optional_yield y)
 {
   bufferlist aclbl;
 
   acls = acl;
   acl.encode(aclbl);
 
-  return store->ctl()->bucket->set_acl(acl.get_owner(), ent.bucket, info, aclbl, null_yield);
+  return store->ctl()->bucket->set_acl(acl.get_owner(), info.bucket, info, aclbl, null_yield);
 }
 
-RGWUser *RGWRadosStore::get_user(const rgw_user &u)
+RGWObject* RGWRadosBucket::get_object(const rgw_obj_key& k)
+{
+  return new RGWRadosObject(this->store, k, this);
+}
+
+RGWUser* RGWRadosStore::get_user(const rgw_user &u)
 {
   return new RGWRadosUser(this, u);
 }
 
+RGWObject* RGWRadosStore::dup_object(RGWObject* _o)
+{
+  RGWRadosObject* obj = static_cast<RGWRadosObject *>(_o);
```

Comment:
this is error-prone since the compiler will accept any kind of RGWObject. consider making `virtual std::unique_ptr<RGWObject> clone() = 0` part of the RGWObject interface instead